backend/src/apiserver/webhook/pipelineversion_webhook.go

```diff
@@ -0,0 +1,135 @@
+	ctrladmission "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	err := k8sapi.AddToScheme(scheme)
+	if err != nil {
+		// Panic is okay here because it means there's a code issue and so the package shouldn't initialize.
+		panic(fmt.Sprintf("Failed to initialize the Kubernetes API scheme: %v", err))
+	}
+}
+
+type PipelineVersionsWebhook struct {
+	Client ctrlclient.Client
+}
+
+var _ ctrladmission.CustomValidator = &PipelineVersionsWebhook{}
+
+func newBadRequestError(msg string) *apierrors.StatusError {
+	return &apierrors.StatusError{
+		ErrStatus: metav1.Status{
+			Code:    http.StatusBadRequest,
+			Reason:  metav1.StatusReasonBadRequest,
+			Message: msg,
+		},
+	}
+}
+
+func (p *PipelineVersionsWebhook) ValidateCreate(
+	ctx context.Context, obj runtime.Object,
+) (warnings ctrladmission.Warnings, err error) {
+	pipelineVersion, ok := obj.(*k8sapi.PipelineVersion)
+	if !ok {
+		return nil, newBadRequestError(fmt.Sprintf("Expected a PipelineVersion object but got %T", pipelineVersion))
+	}
+
+	pipeline := &k8sapi.Pipeline{}
+
+	err = p.Client.Get(
+		ctx, types.NamespacedName{Namespace: pipelineVersion.Namespace, Name: pipelineVersion.Spec.PipelineName}, pipeline,
+	)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, newBadRequestError("The spec.pipelineName doesn't map to an existing Pipeline object")
+		}
+
+		return nil, err
+	}
+
+	pipelineSpec, err := json.Marshal(pipelineVersion.Spec.PipelineSpec.Value)
+	if err != nil {
+		return nil, newBadRequestError(fmt.Sprintf("The pipeline spec is invalid JSON: %v", err))
+	}
+
+	tmpl, err := template.NewV2SpecTemplate(pipelineSpec)
+	if err != nil {
+		return nil, newBadRequestError(fmt.Sprintf("The pipeline spec is invalid: %v", err))
+	}
+
+	err = common.ValidatePipelineName(tmpl.V2PipelineName())
+	if err != nil {
+		return nil, newBadRequestError(err.Error())
+	}
+
+	if pipelineVersion.Name != tmpl.V2PipelineName() {
+		return nil, newBadRequestError("The object name must match spec.pipelineSpec.pipelineInformation.name")
+	}
+
+	return nil, nil
+}
+
+func (p *PipelineVersionsWebhook) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (ctrladmission.Warnings, error) {
+	oldPipelineVersion, ok := oldObj.(*k8sapi.PipelineVersion)
+	if !ok {
+		return nil, newBadRequestError(fmt.Sprintf("Expected a PipelineVersion but got %T", oldObj))
+	}
+
+	newPipelineVersion, ok := newObj.(*k8sapi.PipelineVersion)
+	if !ok {
+		return nil, newBadRequestError(fmt.Sprintf("Expected a PipelineVersion but got %T", newObj))
+	}
+
+	if oldPipelineVersion.Generation != newPipelineVersion.Generation {
+		return nil, newBadRequestError("Pipeline spec is immutable; only metadata changes (labels/annotations) are allowed")
+	}
+
+	return nil, nil
+}
+
+func (p *PipelineVersionsWebhook) ValidateDelete(_ context.Context, _ runtime.Object) (ctrladmission.Warnings, error) {
+	return nil, nil
```

Comment:
I'm guessing this is because we don't do any delete validation, still, it's a bit odd seeing this, maybe a quick comment is worthwhile